### PR TITLE
fix reading of string type param

### LINKED ISSUE #16: Problem with std::string parameter input
Good evening, I am trying to run a calculation using the latest versions on the master branch of AlpsCore and Alpcore/CT-Hyb, and I guess an error, which seems to have to do with the parameters type. An excerpt from my input.ini file is as follows:

```[measurement.equal_time_G1]
on=0

[measurement.two_time_G2]
on=0
n_legendre=50

[measurement.equal_time_G2]
on=0

[measurement.nn_corr]
on=1
n_tau=3
n_def=10
def="nn_def.txt" 
```

When executing hybmat, I get the following error:
```
We cannot open nn_def.txt (type: std::string) (name='measurement.nn_corr.def')!
```
thrown at line 423 of impurity_init.ipp. The cause of the error is that the value of par[fname_key] is actually the string "nn_def.txt (type: std::string) (name='measurement.nn_corr.def')" instead of the string "nn_def.txt". I guess this is coming from ALPSCore, but I am not that experienced with the inner workings of the parameters type in order to identify the root cause.

Thanks!

I guess

### PR DESCRIPTION
**Pull request by dombrno**
fixes #16 

The change in main.cpp is not necessary with respect to any bug I have encountered, but I thought it could not hurt. Please feel free to reject if inappropriate.

### CODE REVIEW

File: src/impurity_init.ipp [hunk 1]
```diff
@@ -362,7 +362,7 @@ void HybridizationSimulation<IMP_MODEL>::read_eq_time_two_particle_greens_meas()
     return;
   }
 
-  std::ifstream infile_f(boost::lexical_cast<std::string>(par[fname_key]).c_str());
+  std::ifstream infile_f(boost::lexical_cast<std::string>(par[fname_key].template as<std::string>()));
```
Comment (galexv) [inline src/impurity_init.ipp]: `lexical_cast` is not needed here (and actually was erroneous, to begin with!). `boost::lexical_cast<string>` works by outputting its argument into a stream; however, there is no guarantee that the resulting string is a value of the argument in any sense: it's just its string representation. (That is, conversion from string representation back to the value may not be correct or at all possible; in this case, a parameter is represented as a string containing, for debugging purposes, its name, value and type.  I am going to change this, see  issue https://github.com/ALPSCore/ALPSCore/issues/490). This code was working with a previous version of parameters accidentally. What is needed here not a lexical cast, but a _static cast_: 

Correct code could be either 
```c++
std::ifstream infile_f(par[fname_key].as<std::string>());
```
or 
```c++
std::ifstream infile_f(static_cast<std::string>(par[fname_key]));
```
or simply
```c++
std::ifstream infile_f(std::string(par[fname_key]));
```

(i do not think that `template` is needed here).

File: src/impurity_init.ipp [hunk 2]
```diff
@@ -417,7 +417,7 @@ void HybridizationSimulation<IMP_MODEL>::read_two_time_correlation_functions() {
...
```
Comment (galexv) [inline src/impurity_init.ipp]: The same comment as above: the lexical cast was erroneous here; it is static cast that is needed:
```c++
std::ifstream infile_f(par[fname_key].as<std::string>());
```
or
```c++
std::ifstream infile_f(std::string(par[fname_key]));
```

File: src/main.cpp [hunk 3]
```diff
@@ -48,7 +48,7 @@ int main(int argc, const char *argv[]) {
 
   //write the results into a hdf5 file
   if (c.rank() == 0) {
-    std::string output_file = par["outputfile"];
+       std::string output_file = par["outputfile"].template as<std::string>();
```
Comment (galexv) [inline src/main.cpp]: In this case, `par["outputfile"]` should unambiguously be converted to `std::string` automatically, so `as<std::string>()` is unnecessary (and i believe `template` is not needed either).

Comment (dombrno) [reply src/main.cpp]: I understand. I will modify following your recommendations and re-submit. Thanks

**Comment by shinaoka** [comment]
Thank you for the contribution and comments!
It seems that the use of as() and static_cast looks correct.
Dominique, could you resubmit changes?

BTW, I need to fix Travis CI...

**Comment by dombrno** [comment]
Dear Hiroshi,
Yes, it will be my pleasure to resubmit the changes, some time today.
I was wondering if I needed to fix anything on my side for the Travis problems, but it looks like not?
Dominique

**Comment by dombrno** [comment]
I think the pull request is ready for review now. Thanks!

**Comment by shinaoka** [comment]
Thank you for the update.

This is just because I have not had time to configure Travis properly... 
I will marge the changes.

File: src/impurity_init.ipp [hunk 4]
```diff
@@ -362,7 +362,7 @@ void HybridizationSimulation<IMP_MODEL>::read_eq_time_two_particle_greens_meas()
     return;
   }
 
-  std::ifstream infile_f(boost::lexical_cast<std::string>(par[fname_key]).c_str());
+  std::ifstream infile_f(par[fname_key].as<std::string>());
```
Comment (galexv) [inline src/impurity_init.ipp]: Apparently i was wrong in my previous comments, and at least clang does require the `template` keyword here (see https://travis-ci.org/ALPSCore/CT-HYB/jobs/355893268#L5641) --- my apologies. Oh well, we'll fix this...

**Comment by dombrno** [comment]
ah clang and its surprises :)

**Comment by shinaoka** [comment]
Oh, it is nice to have Travis CI...

**Pull request merged**